Assets/Features/Building/Scripts/Grid/Cursor.cs

```diff
@@ -0,0 +1,367 @@
+        {
+            Hover(position, _shape);
+
+            if (Input.GetMouseButton(0))
+            {
+                _grid.Set(position, _selectedBuilding.BuildItems[0].Tile);
+            }
+            else if (Input.GetMouseButton(1))
+            {
+                //remove the tiles in the shape of the selection
+                _grid.Remove(position);
+            }
+        }
+
+        #endregion
+
+        private void Hover(Vector3Int position, List<SubBuildItem> shapes, bool requireAllAvailable = false, bool flipColors = false)
+        {
+            //Sets the offset of the whole grid
+            Vector3Int offset = new Vector3Int(Mathf.RoundToInt(_tilemap.transform.position.x),
+                Mathf.RoundToInt(_tilemap.transform.position.y),
+                Mathf.RoundToInt(_tilemap.transform.position.z));
+
+            //get the current selection of tiles
+            List<SubBuildItem> currentSelectedGroup = new List<SubBuildItem>();
+            foreach (var shape in shapes)
+            {
+                currentSelectedGroup.Add(new SubBuildItem(shape.Tile, position + shape.GridPosition));
+            }
+
+            //remove the tiles that were in the previous selection, but not in the current
+            foreach (var gridPosition in _selectedGroup)
+            {
+                if (!currentSelectedGroup.Any(x => x.GridPosition == gridPosition.GridPosition))
+                {
+                    _tilemap.SetTile(gridPosition.GridPosition - offset, null);
+                }
+            }
+
+            //check if all positions are required when placing
+            bool valid = true;
+            if (requireAllAvailable)
+            {
+                foreach (var gridPosition in currentSelectedGroup)
+                {
+                    if (!_grid.IsEmpty(gridPosition.GridPosition))
+                    {
+                        valid = false;
+                        break;
+                    }
+                }
+            }
+
+            var validColor = !flipColors ? _validColor : _invalidColor;
+            var invalidColor = !flipColors ? _invalidColor : _validColor;
+            
+            //set the tile on the tilemap
+            foreach (var gridPosition in currentSelectedGroup)
+            {
+                TileChangeData tempTile = new TileChangeData()
+                {
+                    position = gridPosition.GridPosition - offset,
+                    color = valid
+                        ? (_grid.IsEmpty(gridPosition.GridPosition) ? validColor : invalidColor)
+                        : invalidColor
+                };
+
+                tempTile.transform = Matrix4x4.Rotate(Quaternion.Euler(0, 0, _rotation * -90));
+                tempTile.tile = gridPosition.Tile;
+                _tilemap.SetTile(tempTile, true);
+            }
+
+            _selectedGroup = currentSelectedGroup;
+        }
+
+        public void ChangeToBuilding(BuildableObject buildableObject)
```

Comment:
`ChangeToBuilding()` might not be the right name for this function since we're not changing to a building, instead changing the object to build. Something like `ChangeSelectedBuildable()` might be more suitable.